### PR TITLE
Fixed #7333 - Unminify JavaScript in the SuiteCRM repository

### LINKED ISSUE #7333: Unminify JavaScript in the SuiteCRM repository
Earlier today I opened #7332 to fix some minified JavaScript that was in the FP_Events module, and that made me curious to see if there were any other files that were commited to the git repo with minification. It looks like there are quite a few.

A lot are in the `include/` directory, and I won't include them in this list just because them being minified _sort of_ makes sense (we should really still fix it, though).

For example
- [ ] Most of the JS files in `install/`, e.g. `install/license.js`, `install/oc_install.js`, `install/site_config.js` etc. Though there are corresponding source files in `jssource/src_files/install` it looks like? So maybe that's fine.
- [ ] [`modules/calendar/Cal.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/calendar/Cal.js)
- [ ] [`modules/ACLRoles/ACLRoles.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/ACLRoles/ACLRoles.js)
- [ ] [`modules/AM_ProjectTemplates/jsclass_scheduler.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/AM_ProjectTemplates/jsclass_scheduler.js)
- [ ] [`modules/Campaigns/DetailView.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Campaigns/DetailView.js)
- [ ] [`modules/Campaigns/WebToLead.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Campaigns/WebToLead.js)
- [ ] [`modules/Campaigns/wizard.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Campaigns/wizard.js)
- [ ] [`modules/Connectors/Connector.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Connectors/Connector.js)
- [ ] [`modules/Contacts/Contact.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Contacts/Contact.js)
- [ ] [`modules/Documents/documents.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Documents/documents.js)
- [ ] [`modules/EAPM/EAPMEdit.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/EAPM/EAPMEdit.js)
- [ ] [`modules/EmailTemplates/EmailTemplate.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/EmailTemplates/EmailTemplate.js)
- [ ] [`modules/FP_events/duration_dependency.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/FP_events/duration_dependency.js`) (#7332)
- [ ] [`modules/FP_events/jsclass_scheduler.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/FP_events/jsclass_scheduler.js`) (#7332)
- [ ] [`modules/Meetings/duration_dependency.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Meetings/duration_dependency.js)
- [ ] [`modules/Meetings/jsclass_scheduler.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Meetings/jsclass_scheduler.js)
- [ ] [`modules/MergeRecords/Merge.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/MergeRecords/Merge.js)
- [ ] [`modules/Leads/Lead.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Leads/Lead.js)
- [ ] [`modules/InboundEmail/InboundEmail.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/InboundEmail/InboundEmail.js)
- [ ] [`modules/Home/tour.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Home/tour.js)
- [ ] [`modules/Project/jsclass_scheduler.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Project/jsclass_scheduler.js)
- [ ] [`modules/Schedulers/Schedulers.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Schedulers/Schedulers.js)
- [ ] [`modules/Studio/studio.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Studio/studio.js)
- [ ] [`modules/Studio/studiodd.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Studio/studiodd.js)
- [ ] [`modules/Studio/studiotabgroups.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Studio/studiotabgroups.js)
- [ ] [`modules/Studio/ygDDListStudio.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Studio/ygDDListStudio.js)
- [ ] [`modules/UpgradeWizard/upgradeWizard.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/UpgradeWizard/upgradeWizard.js)
- [ ] [`modules/Users/DetailView.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Users/DetailView.js)
- [ ] [`modules/Users/login.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Users/login.js)
- [ ] [`modules/Users/PasswordRequirementBox.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Users/PasswordRequirementBox.js)
- [ ] [`modules/Users/UserEditView.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Users/UserEditView.js)
- [ ] [`modules/Users/User.js`](https://github.com/salesagility/SuiteCRM/blob/develop/modules/Users/User.js)
- [ ] [`service/utils/SugarRest.js`](https://github.com/salesagility/SuiteCRM/blob/develop/service/utils/SugarRest.js)

Note: I found these by searching for `var` in VS Code and then looking through the results for lines that looked like `var this;var that;var them;var spaghetti;`. This list probably isn't comprehensive.

Most (all?) of these are from the 'Initial Commit' to the repo 6 years ago. I'm not sure if the original SugarCRM files exist somewhere that we can use, or if they were minified in the Sugar repo as well. Assuming we can't recover them from a copy of the Sugar CE repo, the best solution would probably be to do what I did in #7332 and just reformat the files using Prettier.

### PR DESCRIPTION
**Pull request by Dillon-Brown**
## Description

Ideally, we shouldn't be including any autogenerated files with git. This PR removes the minified JS files that are built when running JS repair.

TODO: 
- [X] Create robo task for minifying JS.
- [x] Fix Travis.

## Motivation and Context

Issue reference: #7333

## How To Test This

1. Run ```./vendor/bin/robo build:js```
2. Install CRM and check that the removed files have been rebuilt.

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [X] My code follows the code style of this project found [here](https://docs.suitecrm.com/community/contributing-code/coding-standards/).
- [x] My change requires a change to the documentation.
- [X] I have read the [**How to Contribute**](https://docs.suitecrm.com/community/contributing-code/) guidelines.

### CODE REVIEW

**Comment by Dillon-Brown** [comment]
Confirmed that running RepairJSFile rebuilds these.

[deletedfiles.txt](https://github.com/salesagility/SuiteCRM/files/3261615/deletedfiles.txt)
[newfiles.txt](https://github.com/salesagility/SuiteCRM/files/3261617/newfiles.txt)

The only diff is two JS files that aren't already included, SugarEmailAddressLegacy.js, about.js and some yui3 cache files.

**Comment by connorshea** [comment]
It seems like a lot of files don't actually have counterparts in the jssource directory.

Notably, these are missing:
- `jssource/src_files/modules/Calendar/fullcalendar/lang/*.js`, all the language files in the fullcalendar directory.
- `jssource/src_files/include/SuiteGraphs/rgraph/libraries/*.js`, various libraries for rgraph
- `jssource/src_files/include/javascript/mozaik/vendor/**/*.js`, various libraries for mozaik, e.g. tinymce.
- `jssource/src_files/modules/jjwg_Maps/javascript/*.js`, various files for the jjwg_Maps module
- `jssource/src_files/include/javascript/jstree/**/*.js`, files for the jstree library

And it looks like yui3 is also missing a counterpart source file.

**Comment by connorshea** [comment]
It looks like JSGroupings.php will need to have the file paths for the JavaScript updated: https://github.com/salesagility/SuiteCRM/blob/ef1b41965c056937894dd3018d3b149c9c25a9a9/jssource/JSGroupings.php

**Comment by connorshea** [comment]
>Confirmed that running RepairJSFile rebuilds these.
>
>deletedfiles.txt
>newfiles.txt
>
>The only diff is two JS files that aren't already included, SugarEmailAddressLegacy.js, about.js and some yui3 cache files.

Oh, hm. Somehow I missed this. I wonder why it isn't working in Travis, does the install suite need these files generated before it'll work?

**Comment by Dillon-Brown** [comment]
@connorshea Yeah we'll need these files to be generated before install. This PR should work for manually installing but will need to get them built in Travis too before it will pass.

**Comment by connorshea** [comment]
Conflicts :(

**Comment by Dillon-Brown** [comment]
Should be something related to how Travis is installing the CRM. If you pull this branch you should be able to install as normal without any issues.

**Comment by connorshea** [comment]
So apparently there are a bunch of files in `include/javascript` that are minified and don't have corresponding source files :')

For example, `include/javascript/jquery/` exists (seemingly, all minified), but no corresponding directory exists for these files in `jssource/src_files/include/javascript` :(

This is also true of a bunch of other files.

This isn't worth fixing right now, I'd prefer to have this PR merged without worrying about that, I just wanted to mention it since I kept getting tripped up by it.

File: lib/Robo/Plugin/Commands/BuildCommands.php [hunk 1]
```diff
@@ -49,6 +49,14 @@ class BuildCommands extends \Robo\Tasks
     use RoboTrait;
     // define public methods as commands
 
+    public function buildJS()
```
Comment (connorshea) [inline lib/Robo/Plugin/Commands/BuildCommands.php]: This should have a PHPDoc explaining what it does.

**Comment by cameronblaikie** [comment]
Assessed :+1:

**Comment by connorshea** [comment]
@Dillon-Brown Actually, I want to ask a couple questions after trying this out for a bit:

- This causes 612 files to be created when run, and they're not ignored by git. Should we ignore all `.js` files outside `jssource/`? Otherwise these files would probably get regenerated again.
- Is the `js:build` task run whenever 'Repair and Rebuild All' is run? If not, should it be?
- I wonder if we should have a way to delete the minified files so that they don't stay around forever if the source file is deleted? :?
- It's way too much work right now and would probably break way too many things, but should we be putting the minified JS files into `cache/` instead of the corresponding directories?

**Comment by Dillon-Brown** [comment]
@connorshea 

1. Agreed, although we should consider that there are also some non-minified files that sit outside of jssource as well.
2. The "Repair and Rebuild" in admin panel? I'm not sure it would make sense to add there as you can already do this by running "Rebuild Minified JS Files". Also worth considering that it would greatly increase the time it would take to R&R.
3. Agreed, not sure how you would detect this atm though with some JS files existing only outside of jssource.
4. Would indeed break everything but I agree it would be really nice to have, it would make 3. much easier too.

**Comment by connorshea** [comment]
@Dillon-Brown for #1, any JS files outside jssource are already in the repo, so ignoring them wouldn't change anything (changes to those files would still get picked up). Any new files that are added would have to use the `--force` flag on `git add`, but I don't think we should really add new JS files outside jssource anyway.

I think this should be fixed before this PR can be merged :)

An alternative approach would be to output the files names like `x.min.js` rather than `x.js`, that way we can just ignore `*.min.js`. This approach would still be a pretty big break change, though, so it's probably not worth doing.

**Comment by pgorod** [comment]
Removing 613 files would be a relevant improvement for the Upgrade Wizard problems #8257

**Comment by connorshea** [comment]
FYI you can find a lot of the places where the JS files are imported by searching with this regex: `'file' => '.*.js'`

It probably misses some, but 🤷‍♂

**Comment by connorshea** [comment]
Another thing you can use to find them is searching for uses of `sugar_getjspath`

**Pull request closed without merging**